### PR TITLE
Fix to #7087 - x86 RyuJIT JitstressRegs=2 assert failure

### LINKED ISSUE #7087: RyuJIT x86: Assertion failed '(farthestRefPhysRegRecord != nullptr) && (farthestLocation > refLocation || refPosition->isFixedRegRef)
In Windows_NT.ryujit.x86.Checked.jit_stress_regs.2

set COMPlus_AltJit=*
set COMPlus_AltJitName=protojit.dll
set COMPlus_JitStressRegs=2

/JIT/Regression/VS-ia64-JIT/M00/b113493/b113493/b113493.cmd

> D:\j\workspace\x86_ryujit_ch---eaa82168\bin\tests\Windows_NT.x86.Checked\TestWrappers\JIT\JIT.XUnitWrapper.cs(377535): error : JIT._Regression_VS_ia64_JIT_M00_b113493_b113493_b113493_._Regression_VS_ia64_JIT_M00_b113493_b113493_b113493_cmd [FAIL] [D:\j\workspace\x86_ryujit_ch---eaa82168\tests\runtest.proj]
> Assert failure(PID 3376 [0x00000d30], Thread: 4968 [0x1368]): Assertion failed '(farthestRefPhysRegRecord != nullptr) && (farthestLocation > refLocation || refPosition->isFixedRegRef)' in 'DefaultNamespace.ExternalException:runtest():this' (IL size 345)
> File: d:\j\workspace\x86_ryujit_ch---eaa82168\src\jit\lsra.cpp Line: 5582
> Image: D:\j\workspace\x86_ryujit_ch---eaa82168\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\CoreRun.exe
> Return code: 1
> Raw output file: D:\j\workspace\x86_ryujit_ch---eaa82168\bin\tests\Windows_NT.x86.Checked\Reports\JIT\Regression\VS-ia64-JIT\M00\b113493\b113493\b113493.output.txt
> Raw output:
> BEGIN EXECUTION
> "D:\j\workspace\x86_ryujit_ch---eaa82168\tests\..\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\corerun.exe" b113493.exe
> Expected: 100
> Actual: 123456789
> END EXECUTION - FAILED
> FAILED
> Test Harness Exitcode is : 1

### PR DESCRIPTION
**Pull request by sivarv**
This is a jitstressregs=2 only issue on x86.

IR has the following expression
v01 = v02/(v02-v01)

t0 = GT_SUB(v02, v01)
v01 = GT_DIV(v02, t0)

v02 was allocated to edx and v01 was allocated to ecx.

The Def position of GT_SUB has the following Candidates { ecx, ebx, esi, edi }. Note that eax and edx were not present in Candidates of GT_SUB since those regs are killed by GT_DIV.

Lsra is in allocateBusyReg() for Def position of GT_SUB.
JitstressRegs=2 means LSRA_LIMIT_CALLER. Under this mode, Candidates of GT_SUB get limited to RBM_CALLEE_TRASH (= {eax, ecx, edx} on x86) by intersecting with Candidate set. Note that ecx is the only common between RBM_CALLEE_TRASH and Candidates of GT_SUB. But ecx cannot be allocated to Def position of GT_SUB since ecx is allocated v01 (in GT_SUB Def = v02-v01) is marked as delayRegFree (because targetReg of non-commutative oper like GT_SUB cannot be same as op2's reg).

Fix: StressLimitRegs() now ensures there there are minimum 2 callee trash registers under jitstressregs=2 mode.

Fix #7087

### CODE REVIEW

**Comment by sivarv** [comment]
@CarolEidt - Please review this.
CC @dotnet/jit-contrib

File: src/jit/lsra.cpp [hunk 1]
```diff
@@ -384,7 +384,36 @@ regMaskTP LinearScan::stressLimitRegs(RefPosition* refPosition, regMaskTP mask)
...
+                    // since v01 is marked as delayRegFree. Because targetReg of
+                    // non-commutative opers like GT_SUB cannot be the same as
+                    // op2's reg on xarch.
+                    //
+                    // On x86 alone this needs to be ensured because GT_DIV
+                    // kills two callee trash registers (eax and edx) and op2
+                    // of GT_SUB could take ecx leaving no registers for 
+                    // allocation.  On targets like amd64 this is not an issue
+                    // because there are more callee trash registers leaving
+                    // aside { eax, edx, ecx }
```
Comment (CarolEidt) [inline src/jit/lsra.cpp]: You might want to add a final note to the effect that this code only applies in the JitStressRegs case.

**Comment by CarolEidt** [comment]
LGTM

---
In reply to: [245765061](https://github.com/dotnet/coreclr/pull/7111#issuecomment-245765061) [](ancestors = 245765061)

**Comment by CarolEidt** [comment]
:shipit:

**Comment by sivarv** [comment]
@dotnet-bot retest this please

**Pull request merged**